### DISCUSSION TITLE
S3CopyObjectOperator when used within a dynamically mapped task group will use entire list of dynamically mapped input within `s3_key` and `data`.

### DISCUSSION THREAD

**Discussion by TJaniF** [apache/airflow#29383] [Q&A] [answered]
### Apache Airflow Provider(s)

amazon

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==7.2.0

### Apache Airflow version

2.5.1

### Operating System

macOS 13.1 (22C65)

### Deployment

Other Docker-based deployment

### Deployment details

Astro CLI (astro-runtime:7.2.0) 

### What happened

When dynamically mapping over a task group containing the S3CopyObjectOperator and using the mapped value within the `s3_key` or `data` parameter of the operator; the operator will use the full list of dynamically mapped inputs in every mapped task group instance. 

Using the following mapped task group:

```python
    @task_group()
    def create_s3_files(num):

        @task
        def return_num_as_int(my_num):
            return my_num

        my_num_as_int = return_num_as_int(num)

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data=json.dumps(f"{my_num_as_int}"),
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/{my_num_as_int}.txt"
        )

        my_num_as_int >> write_to_s3

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])
```

Creates 6 mapped task group instances with 2 tasks each as expected. The `return_num_as_int` task returns the individual inputs for `num` as expected. XCom for map index 3 in the screenshot:

<img width="1142" alt="Screenshot 2023-01-30 at 15 38 39" src="https://user-images.githubusercontent.com/90063506/215507228-6055b2ba-8b31-4f0c-90bb-56f827703424.png">

The `write_to_S3` task will render the full list that is being mapped over for everyone of its six instances within the mapped task group as shown in the screenshot of the Rendered Template:

<img width="758" alt="Screenshot 2023-01-30 at 15 43 06" src="https://user-images.githubusercontent.com/90063506/215508388-b1e4f275-1890-4259-9bc3-82d815be7ffd.png">

The file created in the S3 bucket matches the rendered template with the name `[0, 1, 2, 3, 4, 5].txt` and the content `"[0, 1, 2, 3, 4, 5]"`. 

### Variations

**Changing the replace parameter**

When setting `replace=True` in the operator, one file is created, all task instances within the mapped task group are marked as successful. When setting `replace=False` 1 mapped task instance is successful creating the same file with the same content with the other mapped task instances failing with the following error:

```
[2023-01-30, 14:46:35 UTC] {taskinstance.py:1768} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/providers/amazon/aws/operators/s3.py", line 399, in execute
    s3_hook.load_string(
  File "/usr/local/lib/python3.9/site-packages/airflow/providers/amazon/aws/hooks/s3.py", line 99, in wrapper
    return func(*bound_args.args, **bound_args.kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/providers/amazon/aws/hooks/s3.py", line 70, in wrapper
    return func(*bound_args.args, **bound_args.kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/providers/amazon/aws/hooks/s3.py", line 695, in load_string
    self._upload_file_obj(file_obj, key, bucket_name, replace, encrypt, acl_policy)
  File "/usr/local/lib/python3.9/site-packages/airflow/providers/amazon/aws/hooks/s3.py", line 765, in _upload_file_obj
    raise ValueError(f"The key {key} already exists.")
ValueError: The key [0, 1, 2, 3, 4, 5].txt already exists.
```

**Using the mapped input directly**

When using the mapped input directly in the S3CopyObjectOperator parameter as such:

```python
    @task_group()
    def create_s3_files(num):

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data=json.dumps(f"{num}"),
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/{num}.txt"
        )

        write_to_s3

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])
```

The resulting Rendered Template uses the full MappedArgument (this matches the one file created in S3):

<img width="758" alt="Screenshot 2023-01-30 at 15 51 05" src="https://user-images.githubusercontent.com/90063506/215510413-d700998e-653b-44ba-8d49-8ccc95907d47.png">

**Mapping over the operator directly works as expected**

When mapping over the S3CopyObjectOperator directly, both within or outside of a task group. The results are as expected one file created for each mapped task instance. 

```python
    @task_group()
    def create_s3_files(list_of_keys):

        write_to_s3 = S3CreateObjectOperator.partial(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data=json.dumps("hi"),
            replace=True,
        ).expand(s3_key=list_of_keys)

    create_s3_files(list_of_keys=[f"{MY_S3_BUCKET}/0.txt",f"{MY_S3_BUCKET}/1.txt"])
```

The code above creates 2 files. One named `0.txt` the other named `1.txt`.

### What you think should happen instead

The S3CopyObjectOperator should create one file in S3 for each element in the input that the task group is being mapped over only using the element with the correct map index. 

Instead of creating `[0, 1, 2, 3, 4, 5].txt` six times I would expect 6 files to be created named `0.txt`, `1.txt`, `2.txt` etc... 

This behavior would match the behavior of the task flow task within the dyncamically matched task group which has the outputs `0`, `1`, `2` etc for each of its task instances within the mapped task group instances. It would also match the behavior seen when mapping over the S3CopyObjectOperator directly.

### How to reproduce

Full DAG to reproduce. Needs 1 S3 bucket and an AWS conn id. 

Reproduction steps:

1) Add a valid `AWS_CONN_ID` and `MY_S3_BUCKET` after the package imports.
2) Run the DAG.
3) View the file created in the S3 bucket.

```python
from airflow.decorators import dag, task_group, task
from airflow.providers.amazon.aws.operators.s3 import (
    S3CreateObjectOperator
)
from pendulum import datetime

import json

MY_S3_BUCKET = "s3://mytxtbucket"
AWS_CONN_ID = "aws_conn"

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def S3_filename_bug_dag():

    @task_group()
    def create_s3_files(num):

        @task
        def return_num_as_int(my_num):
            return my_num

        my_num_as_int = return_num_as_int(num)

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data=json.dumps(f"{my_num_as_int}"),
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/{my_num_as_int}.txt"
        )

        my_num_as_int >> write_to_s3

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])


S3_filename_bug_dag()
```

### Anything else

cc: @kentdanas 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
Did you tried [Dynamic Task Mapping: Mixing TaskFlow and classic operators](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html#mixing-taskflow-and-classic-operators)? Does it work as expected in this case or also create not expected output/maps?

**Comment by TJaniF** [comment]
Hi @Taragolis 

Yes, using the S3Hook directly like in the Mixing TaskFlow and classic operators works as expected and creates 6 separate files:

```python
from airflow.decorators import dag, task_group, task
from airflow.providers.amazon.aws.hooks.s3 import S3Hook
from pendulum import datetime
import io
import json

MY_S3_BUCKET = "s3://mytxtbucket"
AWS_CONN_ID = "aws_conn"

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def mixed():

    @task_group()
    def create_s3_files(num):

        @task
        def create_file(aws_conn_id, bucket, num):
            hook = S3Hook(aws_conn_id=aws_conn_id)

            return hook.load_file_obj(file_obj=io.BytesIO(b'something'), key=f"{bucket}/x/{num}.txt")
        create_file("aws_conn", MY_S3_BUCKET, num)

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])

mixed()
```

Gives me:

<img width="1233" alt="Screenshot 2023-01-30 at 16 40 27" src="https://user-images.githubusercontent.com/90063506/215522862-d1b0ab2e-48fe-468f-afdf-fdd0c86bd50d.png">

**Comment by josh-fell** [comment]
@TJaniF Another variation, but what happens if you don't use `json.dumps()` on the `XComArg` being passed and without the `.partial()` on the operator? The manipulation of the `XComArg` _might_ be causing an oddity?

```py

    @task_group()
    def create_s3_files(num):

        @task
        def return_num_as_int(my_num):
            return my_num

        my_num_as_int = return_num_as_int(num)

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data="Hi",
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/{my_num_as_int}.txt"
        )

        my_num_as_int >> write_to_s3

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])
```

**Comment by Taragolis** [comment]
@TJaniF  I think it should be something like this

```python
from airflow.decorators import dag, task_group, task
from airflow.providers.amazon.aws.operators.s3 import (
    S3CreateObjectOperator
)
from pendulum import datetime

import json

MY_S3_BUCKET = "s3://mytxtbucket"
AWS_CONN_ID = "aws_conn"

@dag(
    start_date=datetime(2022, 12, 1, tz="UTC"),
    schedule=None,
    catchup=False,
)
def S3_filename_bug_dag():

    @task_group()
    def create_s3_files(num):

        @task
        def return_num_as_int(my_num):
            return my_num

        @task
        def create_copy_kwargs(my_num):
            return {
                "data": json.dumps(my_num),
                "s3_key": f"{MY_S3_BUCKET}/{my_num}.txt",
            }

        my_num_as_int = return_num_as_int(num)
        kwargs = create_copy_kwargs(my_num_as_int)

        S3CreateObjectOperator.partial(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            replace=True
        ).expand_kwargs(kwargs)

    create_s3_files.expand(num=[0,1,2,3,4,5])


S3_filename_bug_dag()

```

Unfortunetly when I tried I got

```console

Broken DAG: [/files/dags/issue_29238.py] Traceback (most recent call last):
  File "<attrs generated init airflow.models.mappedoperator.MappedOperator>", line 27, in __init__
    self.__attrs_post_init__()
  File "/opt/airflow/airflow/models/mappedoperator.py", line 309, in __attrs_post_init__
    raise NotImplementedError("operator expansion in an expanded task group is not yet supported")
NotImplementedError: operator expansion in an expanded task group is not yet supported
```

**Comment by TJaniF** [comment]
Hey @josh-fell 

Passing in `num` directly creates one file with `MappedArgument` in the name and content:
<img width="1172" alt="Screenshot 2023-01-30 at 18 47 38" src="https://user-images.githubusercontent.com/90063506/215554215-42b32b16-4bd6-45ba-b9ea-f06459de7f7a.png">

```python
from airflow.decorators import dag, task_group, task
from airflow.providers.amazon.aws.operators.s3 import (
    S3CreateObjectOperator
)
from pendulum import datetime

import json

MY_S3_BUCKET = "s3://mytxtbucket"
AWS_CONN_ID = "aws_conn"

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def S3_filename_bug_dag_num_direct():

    @task_group()
    def create_s3_files(num):

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data=json.dumps(f"{num}"),
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/{num}.txt"
        )

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])


S3_filename_bug_dag_num_direct()
```

Not using json.dumps ends with the same issue:

<img width="1172" alt="Screenshot 2023-01-30 at 18 50 32" src="https://user-images.githubusercontent.com/90063506/215554835-0eec041a-6e7d-41bc-b0f0-bd76c89b2ea7.png">

for:

```python
from airflow.decorators import dag, task_group, task
from airflow.providers.amazon.aws.operators.s3 import (
    S3CreateObjectOperator
)
from pendulum import datetime


MY_S3_BUCKET = "s3://mytxtbucket"
AWS_CONN_ID = "aws_conn"

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def S3_filename_bug_dag_no_json_dumps():

    @task_group()
    def create_s3_files(num):

        @task
        def return_num_as_int(my_num):
            return my_num

        my_num_as_int = return_num_as_int(num)

        write_to_s3 = S3CreateObjectOperator(
            task_id="write_to_s3",
            aws_conn_id=AWS_CONN_ID,
            data="Hi",
            replace=True,
            s3_key=f"{MY_S3_BUCKET}/a/{my_num_as_int}.txt"
        )

        my_num_as_int >> write_to_s3

    tg_object = create_s3_files.expand(num=[0,1,2,3,4,5])


S3_filename_bug_dag_no_json_dumps()
```

@Taragolis yeah, mapping tasks within mapped task groups seems to be a future thing at this point 😅 . 


**More variations:**

Using a traditional PythonOperator works as expected (each Python task in a mapped task group instance only uses one of the inputs):

```python
from airflow.decorators import dag, task_group, task
from airflow.operators.python import PythonOperator
from pendulum import datetime

def multiply_by_10(number):
    return number * 10

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def python_map(): 

    @task_group()
    def say_things(num):

        @task 
        def turn_num_to_int(num):
            return num 

        int_num = turn_num_to_int(num)

        t1 = PythonOperator(
            task_id="t1",
            python_callable=multiply_by_10,
            op_args=[int_num] 
        )

        int_num >> t1 

    tg_object = say_things.expand(num=[0,1,2,3,4,5])

python_map()
```

But when using `op_args=["{{ ti.xcom_pull(task_ids=['turn_num_to_int'], key='return_value') }}"]` instead I get:

```
[2023-01-30, 18:11:44 UTC] {python.py:177} INFO - Done. Returned value was: [][][][][][][][][][]
```

Switching out the Python for a BashOperator leads to a (on the surface) similar looking error as with the S3CopyObjectOperator:

```python
from airflow.decorators import dag, task_group, task
from airflow.operators.bash import BashOperator
from pendulum import datetime

@dag(
    start_date=datetime(2022, 12, 1),
    schedule=None,
    catchup=False,
)
def bash_map(): 

    @task_group()
    def say_things(num):

        @task 
        def turn_num_to_int(num):
            time.sleep(5)
            return num 

        int_num = turn_num_to_int(num)

        t2 = BashOperator(
            task_id="t2",
            bash_command=f"echo {int_num}"
        )

        int_num >> t2

    tg_object = say_things.expand(num=[0,1,2,3,4,5])


bash_map()
```

Prints:
```
[2023-01-30, 18:15:35 UTC] {subprocess.py:63} INFO - Tmp dir root location: 
 /tmp
[2023-01-30, 18:15:35 UTC] {subprocess.py:75} INFO - Running command: ['/bin/bash', '-c', 'echo [0, 1, 2, 3, 4, 5]']
[2023-01-30, 18:15:35 UTC] {subprocess.py:86} INFO - Output:
[2023-01-30, 18:15:35 UTC] {subprocess.py:93} INFO - [0, 1, 2, 3, 4, 5]
[2023-01-30, 18:15:35 UTC] {subprocess.py:97} INFO - Command exited with return code 0
```

for every map index.

Same when using 

```
bash_command=f"echo $MYNUM",
env={"MYNUM": str(int_num)}
```

in the operator.

So the issue is not contained to the S3CopyObjectOperator but not with all traditional operators. 

Is it possible that the mapped input is evaluated differently when in a Python function (no matter if decorated or traditional PythonOperator than when in an operator input?

**Comment by Taragolis** [comment]
> But when using `op_args=["{{ ti.xcom_pull(task_ids=['turn_num_to_int'], key='return_value') }}"]` instead I get:
> 
> ```
> [2023-01-30, 18:11:44 UTC] {python.py:177} INFO - Done. Returned value was: [][][][][][][][][][]
> ```

I think there is missing `map_indexes` value in `ti.xcom_pull`

**Comment by josh-fell** [comment]
`op_args=["{{ ti.xcom_pull(task_ids=['turn_num_to_int'], key='return_value') }}"]` Will also return a list because the `task_ids` value is a list FYI

**Comment by TJaniF** [comment]
👍 
`op_args=["{{ ti.xcom_pull(task_ids=['turn_num_to_int'], key='return_value') }}"]` 
results in `[][][][][][][][][][]`

`op_args=["{{ ti.xcom_pull(task_ids='turn_num_to_int', key='return_value') }}"]`
results in `NoneNoneNoneNoneNoneNoneNoneNoneNoneNone`

`op_args=["{{ ti.xcom_pull(task_ids='turn_num_to_int', key='return_value', map_indexes=[2]) }}"]`
results in `[][][][][][][][][][]`

`op_args=["{{ ti.xcom_pull(task_ids='turn_num_to_int', key='return_value', map_indexes=2) }}"]`
results in `NoneNoneNoneNoneNoneNoneNoneNoneNoneNone`

🤔

**Comment by josh-fell** [comment]
Ha! It's in a TaskGroup. The `task_id` is "say_things.turn_num_to_int".

**Comment by TJaniF** [comment]
🤦 of course! (I am using this issue as an example in the webinar tomorrow 😂 )

Interesting!

`op_args=["{{ ti.xcom_pull(task_ids='say_things.turn_num_to_int', key='return_value') }}"]`
results in `[0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5][0, 1, 2, 3, 4, 5]`

but `op_args=["{{ ti.xcom_pull(task_ids='say_things.turn_num_to_int', key='return_value', map_indexes=2) }}"]`

results in `2222222222`, so as I would have expected, pulling one mapped task index as a string and multiplying by 10.

For good measure:
`op_args=["{{ ti.xcom_pull(task_ids='say_things.turn_num_to_int', key='return_value', map_indexes=[2,3]) }}"]`
results in `[2, 3][2, 3][2, 3][2, 3][2, 3][2, 3][2, 3][2, 3][2, 3][2, 3]`

**Comment by josh-fell** [comment, marked as answer]
It looks like the `str()` tranformation, and probably any transformation on the `XComArg` causes an issue (at least for this latest example):
```
    @task_group()
    def say_things(num):

        @task
        def turn_num_to_int(num):
            time.sleep(5)
            return num

        int_num = turn_num_to_int(num)

        t2 = BashOperator(
            task_id="t2",
            bash_command="echo $MY_NUM",
            env={"MY_NUM": int_num}
```
Looking at an individual mapped index, the rendered value is what is expected:
<img width="728" alt="image" src="https://user-images.githubusercontent.com/48934154/215568597-dcda35e5-c968-4d68-8b49-f0d89b18ce07.png">
However, using `str(int_num`) instead gives us the entire mapped outputs:
<img width="676" alt="image" src="https://user-images.githubusercontent.com/48934154/215568934-ba8f778c-c0bc-4416-8800-07e66ed256b8.png">
But of course the `BashOperator` task fails because a string argument is expected in the command.

The difference in rendering makes sense since the `PlainXComArg.__str__` is `ti.xcom_pull(task_ids=<task_id>, dag_id=<dag_id>)` which, as @Taragolis alluded to, doesn't consider `map_index`.

**Comment by alexwainger** [comment]
I'm trying to do something similar, i.e. create a dynamic task group from a list of file names, then pass each file name into a traditional operator (S3KeySensor to be specific).

From this thread, it seems like the cleanest way to get this to work is to pass the mapped argument through a taskflow task that simply returns the value, and then pass the output of that task to the traditional operator, is that right? Is there no way to directly pass the mapped argument into a traditional operator inside of a dynamic task group?

cc: @josh-fell @TJaniF

**Reply by alexwainger** [reply]
Actually this doesn't work, the traditional operator ends up getting passed the full list rather than a value from the list as @TJaniF showed in the original post. So what is the recommended way to this?

**Reply by markbegma** [reply]
Same here, i am trying the following thing in my task group:

```
auth_kwargs_list = get_accounts(date="{{ ds }}")

@task_group(group_id="grpid")
def grp(*, text, auth):
    http_op = SimpleHttpOperator(
        task_id="http_task",
        http_conn_id="my_api",
        endpoint=f"/my_endpoint",
        method="GET",
        headers={
            "Content-Type": "application/json",
            "Authorization": auth,
        },
        do_xcom_push=True,
        log_response=True,
    )
    postgres_op = PostgresOperator(
        task_id="postgres_task",
        sql=f"""
            INSERT INTO table
            (text, result)
            VALUES ('{text}', '{http_op.output}')
            """,
        postgres_conn_id="pg_conn",
        database="db",
    )
grp.expand_kwargs(auth_kwargs_list)
```

Quite surprisingly, SimpleHttpOperator expands `auth` just fine, while PostgresOperator struggles with `text`, representing it as:
MappedArgument(_input=ListOfDictsExpandInput(value=XComArg(<Task(_PythonDecoratedOperator): get_accounts>)), _key='text')

So it indeed is some problem with string representation?

**Reply by alexwainger** [reply]
@markbegma the cleanest way I found to make this work is to use the xcom_pull and map_index macros to get the appropriate value out of the list you're mapping over. So your postgres operator would look like this:

```
postgres_op = PostgresOperator(
    task_id="postgres_task",
    sql=f"""
        INSERT INTO table
        (text, result)
        VALUES ('{{ti.xcom_pull(task_ids="auth_kwargs_list")[ti.map_index]}}', '{http_op.output}')
        """,
    postgres_conn_id="pg_conn",
    database="db",
)
```

Or something like that, not exactly sure what the auth_kwargs_list variable looks like.
